### ISSUE TITLE
Review screen-reader usability across the app

### ISSUE THREAD

**Issue by gnprice** [zulip/zulip-flutter#537]
This is a followup to #535, an expanded version to be comprehensive across the app's functionality. This issue is effectively the zulip-flutter successor to the open zulip-mobile issue https://github.com/zulip/zulip-mobile/issues/4394 .

For this issue, the task is to learn to use the screen reader provided by your device's operating system (VoiceOver on iOS, TalkBack on Android), then try to use it with all kinds of features in the app, and see what doesn't work or doesn't work as well as it could.

For everything we find that can be improved, we'll file and triage issues so we can fix them. Use the [a-a11y label](https://github.com/zulip/zulip-flutter/labels/a-a11y) so that it's easy to find the list of them.

### How to approach this issue

See my discussion in https://github.com/zulip/zulip-mobile/issues/4394 about how to start learning to use these systems and to evaluate the app with them.

As part of #535 we'll have covered the essentials of the app's functionality before we start on this followup issue. Other feature areas to cover include:
* reactions
* all kinds of message content features (including interactions like links, images, and spoilers, and Zulip "widgets" like polls)
* the lightbox
* action sheets and their actions
* compose typeaheads
* and everything else in the app.

As part of this issue we'll also be evaluating to higher standards of UX quality than for #535. See https://github.com/zulip/zulip-mobile/issues/4394 for examples of kinds of glitches that can be present, and that we should find and fix:
* properly announcing what page you're on, when you enter a list of items, etc.
* all buttons, even those less commonly used, should work and should have captions
* nothing should be selectable that doesn't actually do anything or give any useful information
* proper conventions for navigating between messages, announcing message sender/time/etc., and so on
* and anything else that comes up that seems like it'd get in the way of using the app's features successfully or efficiently.

For the issues we find as part of this review, generally the fixes will involve Flutter's [`Semantics`](https://main-api.flutter.dev/flutter/widgets/Semantics-class.html) widget and related parts of the Flutter API, to provide the appropriate semantic information for screen readers (like VoiceOver and TalkBack) and other assistive technology to use. But this issue is only about reviewing the app to find those issues; we'll use their own respective issue threads to track actually fixing them.

**Comment by PIG208** [comment]
I spent some time playing around with TalkBack on my Samsung device, on v0.0.33, and wrote down my experience below.

The gestures I used the most are swipe-left and swipe-right, which move the focus indicator. When I mention "before"/"after", I refer to the traversal order of this focus indicator, using these swipe gestures, not the "focus" in Flutter.

The list was originally intended for #535, so it covers some, but not the full breadth of features. We should check the list items when they're filed as an issue or when they're resolved as complete/not planned.

## choose-account page

- [ ] The 3-dots menu comes before the "Add an account" button since it is on the app bar, but its caption is just "show menu". This may be confusing because the menu does not contain essential information, except for the app version and licenses.
- [ ] The "Your Zulip server URL" text field does not show its hint text until it is focused. When it first gains focus, the hint text is not read either.
- [ ] The "----OR-----" separator might need a caption that elaborates what it means. Without a visual cue, it is a bit abrupt as a single word. (Note, by navigating the third-party sign-in sites, we can compare those pages' accessibility with ours quite conveniently.) Trying it on GitHub, there is an invisible element right after the "sign in" button! It reads "password log-in alternatives", and the option after it is "sign in with a passkey".
- [ ] The inputs and their label are read together. The screen reader reads the content before reading the label. On sites like GitHub and GitLab, the label and the input are separate, where the label comes before the input.
- [ ] The loading indicator does not get read, and there is seemingly no way to know if the page is loading.

## Home page

- [ ] The bottom tab navigation buttons do not have labels. The caption for each of them is just: "button, double tap to activate".

## Inbox

- [ ] There is no way to distinguish channel headers and topic items.
- [ ] The extra labels on the topic items might need some captions:
  - [ ] Unread count is read, but it's not mentioned to be the unread count. (E.g.: "some topic name 24");
  - [ ] The topic visibility icons seem to be ignored.
- [ ] The collapse/uncollapse state of the channel headers is not captioned.
- [ ] When collapsing or uncollapsing a channel header, the focus sometimes gets moved elsewhere unexpectedly. -> Fix in #2120
- [ ] Whenever you navigate back from a different page, before moving the blue box to where you left off last time, it is temporarily moved to read the title of the page first. Wonder if we can do the same to topic items, so that when navigating from the message list back to the inbox, it reads "Inbox, channel name, topic name", instead of "Inbox, topic name" (current behavior).

### Topic action sheet

- [ ] The action sheet buttons are not recognized as buttons. For example, "Mute topic" is read as "mute topic, double tap to activate", whereas "Cancel" is read as "Cancel, button, double tap to activate".

## Channels page

- [ ] The screen reader can't tell if a channel is muted.

## Message list

- The traversal order is interesting: App bar -> the sticky recipient header -> whatever message that is at the top of the screen.
- [ ] The unread marker is not read. It's probably a UX problem since there can be a lot of consecutive unreads where it might get too repetitive. 
- [ ] You can focus on the sender row (the UI element that allows you to access their profile) when there are emoji reactions; otherwise, you can't. When you do get access to it, the traversal order is like this: message sent time and its content -> the sender row -> reactions.
- [ ] The traversal order of message content appears inconsistent with some features, especially when there are links:
  - [ ] When there is quoted text and links, the quoted text can get read first, even if it comes after the rest of the message.
  - [ ] When there are both links and mentions, the traversal order within the message seems to be incorrect.
  - [ ] Traversal order of polls is not ideal. The screen reader would read all the poll options at once, but the vote buttons need to be traversed individually.
- [ ] The message list sometimes does not scroll along when the focus moves off-screen.
- [ ] The sticky header can get focused when moving focus backwards. This is unexpected when the user intended to move to the previous message in the same conversation.

### Compose box

- [ ] Autocomplete still works, but it might be helpful to mention when the autocomplete pop-up appears or disappears. The screen reader already does that when the keyboard is shown or hidden.
- [ ] The disabled state of the send button is not read.
